communications/views.py

```diff
@@ -1,12 +1,148 @@
+    pagination_class = None
+    
+    def get_queryset(self):
+        """Return only active notification types."""
+        return NotificationType.objects.filter(is_active=True)
+
+
+
+class UserNotificationPreferenceViewSet(viewsets.ModelViewSet):
+    """
+    API endpoint that allows users to view and update their notification preferences.
+    """
+    serializer_class = UserNotificationPreferenceSerializer
+    permission_classes = [permissions.IsAuthenticated]
+    http_method_names = ['get', 'patch', 'head', 'options']
+
+    def get_queryset(self):
+        """Return only the current user's preferences."""
+        return UserNotificationPreference.objects.filter(user=self.request.user)
+
+    def get_object(self):
+        """
+        Return the current user's preferences, creating them if they don't exist.
+        """
+        queryset = self.filter_queryset(self.get_queryset())
+        obj = queryset.first()
+        
+        if obj is None:
+            obj = UserNotificationPreference.objects.create(user=self.request.user)
+            
+            notification_types = NotificationType.objects.filter(is_active=True)
+            for notification_type in notification_types:
+                UserNotificationTypePreference.objects.create(
+                    user_preference=obj,
+                    notification_type=notification_type,
+                    frequency='immediate'
+                )
+        
+        return obj
+
+    @action(detail=True, methods=['patch'])
+    def update_type_preference(self, request, pk=None):
+        """
+        Update a specific notification type preference.
+        Expected payload: {"notification_type_id": 1, "frequency": "immediate"}
+        """
+        preference = self.get_object()
+        notification_type_id = request.data.get('notification_type_id')
+        frequency = request.data.get('frequency')
+        
+        if not notification_type_id or not frequency:
+            return Response(
+                {'error': 'notification_type_id and frequency are required'},
+                status=status.HTTP_400_BAD_REQUEST
+            )
+        
+        try:
+            type_pref = preference.type_preferences.get(
+                notification_type_id=notification_type_id
+            )
+            type_pref.frequency = frequency
+            type_pref.save()
+            
+            serializer = UserNotificationTypePreferenceSerializer(type_pref)
+            return Response(serializer.data)
+            
+        except UserNotificationTypePreference.DoesNotExist:
+            return Response(
+                {'error': 'Notification type preference not found'},
+                status=status.HTTP_404_NOT_FOUND
+            )
+
+    @action(detail=False, methods=['get'])
+    def options(self, request):
+        """Return available options for notification preferences."""
+        from .models import NotificationFrequency
```

Comment:
The options action overlaps with the default OPTIONS method provided by DRF. Consider renaming this action to avoid confusion with HTTP OPTIONS, or ensure that its purpose is well documented.